cmd/snap-confine/snap-confine-invocation.c

```diff
@@ -78,7 +80,33 @@ void sc_cleanup_invocation(sc_invocation *inv) {
     if (inv != NULL) {
         sc_cleanup_string(&inv->snap_instance);
         sc_cleanup_string(&inv->base_snap_name);
+        sc_cleanup_string(&inv->orig_base_snap_name);
         sc_cleanup_string(&inv->security_tag);
         sc_cleanup_string(&inv->executable);
     }
 }
+
+void sc_maybe_pick_alt_base_snap(sc_invocation *inv) {
+    /* As a special fallback, allow the base snap to degrade from "core" to
+     * "ubuntu-core". This is needed for the migration tests. */
+    char mount_point[PATH_MAX] = {0};
+    sc_must_snprintf(mount_point, sizeof mount_point, "%s/%s/current/", SNAP_MOUNT_DIR, inv->base_snap_name);
+
+    if (access(mount_point, F_OK) == 0) {
+        return;
+    }
+    /* The mount point is missing but for some snaps we may have an alternative. */
+    if (sc_streq(inv->base_snap_name, "core")) {
+        /* For "core" we can still use the ubuntu-core snap. This is helpful in
+         * the migration path when new snap-confine runs before snapd has
+         * finished obtaining the core snap. */
+        sc_must_snprintf(mount_point, sizeof mount_point, "%s/%s/current/", SNAP_MOUNT_DIR, "ubuntu-core");
+        if (access(mount_point, F_OK) == 0) {
+            sc_cleanup_string(&inv->base_snap_name);
+            inv->base_snap_name = sc_strdup("ubuntu-core");
+            debug("falling back to ubuntu-core instead of unavailable core snap");
+            return;
+        }
+    }
+    die("cannot proceed without base snap %s", inv->base_snap_name);
```

Comment:
Hm, I would not expect (from the name) that sc_maybe_pick_alt_base_snap() woudl die() if it cannot find the base snap. Also we also check this elsewhere and die. And in the test the error message comes from the other place it seems. Do we need to die() here? Or am I misreading the code?

Reply:
Indeed, that's unexpected. As if the code didn't run at all. I will debug this.

Reply:
This message is picked up by the test. Everything is fine now.